### DISCUSSION TITLE
Removing Spectral Noise from spectra?

### DISCUSSION THREAD

**Discussion by jwyco12** [thomased/colRverse#12] [Q&A] [answered]
Good day everyone,

I have been collecting spectral samples for my thesis project. On my measurements, I am consistently getting these large peaks that my PI and I believe are electrical noise/artifacts. I was wondering what the optimal method is to deal with the noise. I know about the smoothing function, but was wondering if there is any other method that is more optimal given how large the peaks are. I attached a picture showing the noise, which are the two large peaks as what I am measuring is white plumage. 
![Spec noise](https://github.com/colrverse/colRverse/assets/85466250/038f9508-41cb-46a9-80c0-5fe11cb9c215)

**Comment by thomased** [comment, marked as answer]
Hi Jacob, 

Just copying this here from our email, for posterity. Cheers!

I’ve seen that kind of noise before from light sources (is it a JAZ spec?) – odd. At any rate, with noise that severe I’d suggest  the simplest way to correct for it is just to manually interpolate across the span of the noise (i.e. draw a straight line from just before the noise to just after it), then run it through the smoothing. There’s no way procspec() smoothing will be able to handle it directly, since you’d have to make the smoother so extreme that it’d destroy the shape of the spec itself. Here’s a chunk of code to show what I mean – see how you go, and let me know if you hit any issues!

```
# Library
library(pavo)

# Simulate a simple spectrum
reflect1 <- simulate_spec(wl_inflect = 550)

# Plot it
plot(reflect1)

# Add an ugle 'noise' spike 500-520 nm
reflect1$spec_i550[reflect1$wl >= 500 & reflect1$wl <= 520] <- reflect1$spec_i550[reflect1$wl >= 500 & reflect1$wl <= 520] + 
  rnorm(sum(reflect1$wl >= 500 & reflect1$wl <= 520), mean = 50, sd = 10)

# Plot it
plot(reflect1)

# Remove the noise spike by interpolating a straight line starting either side of the noise
reflect1$spec_i550[reflect1$wl >= 500 & reflect1$wl <= 520] <- 
  approx(c(499, 521), c(reflect1$spec_i550[reflect1$wl == 499], reflect1$spec_i550[reflect1$wl == 521]), xout = reflect1$wl[reflect1$wl >= 500 & reflect1$wl <= 520])$y

# Smooth it as usual using pavo
reflect1_denoise <- procspec(reflect1, opt = 'smooth')

# Plot it
plot(reflect1_denoise)
```
